### PR TITLE
ppl: pass contextual information through policy

### LINKED ISSUE #2609: PPL Contextual Details
Currently `allow` rules in Pomerium Policy Language (PPL) simply return true or false without identifying why they're true or false. This limits our ability to provide better error messages and custom workflows.

**Describe the solution you'd like**

The below description is somewhat confusing since its very technical. Hopefully any PR implementing it will be more clear.

We should update the PPL rego generation of criteria to return an array of values instead of a single boolean. The array will consist of a boolean indicating whether or not the criterion passed as well as a set of strings indicating the reason it passed or failed. The strings will be pre-defined single-word constants that can be used by the Authorize service. For example:

```yaml
user:
  is: "xyz"
```

Might produce:

```rego
user_0 = v {
    v = [true, {"user-ok"}]
}
```

Logical operators will also need to be updated to merge the reason sets accordingly. The final result should be something like:

```json
[true, ["user-ok","claim-ok"]]
```

Or:

```json
[false, ["user-unauthorized","claim-unauthorized"]]
```

Theoretically this also gives us the ability to distinguish between `user-unauthorized` and `user-unauthenticated`. Currently we handle this in Authorize directly to cause a redirect to Authenticate when a user hasn't logged in yet, but with this update we could simply look for the `user-unauthenticated` reason when `allow` is `false`.

In addition to updating the criteria we will need to add a new rego function to convert any existing return values into the new form, in case additional criteria conform to the old way of doing things.

**Describe alternatives you've considered**

Another option might be to use a global variable in the rego to represent these values. But rego makes it difficult to update such a global variable since its not a traditional imperative programming language.

### PR DESCRIPTION
**Pull request by calebdoxsey**
## Summary
Update Pomerium Policy Language (PPL) rego generation to support bubbling up contextual information through the policy evaluation. Previously results were simple bools. They are now `[boolean, set<string>]` tuples. Using this contextual information we can customize error messages and implement custom behavior.

There are three special cases we currently handle:

1. If `route-not-found` is returned, a 404 is issued.
2. If `invalid-client-certificate` is returned, a 495 is issued.
3. If `user-unauthenticated` is returned, a redirect to Authenticate is issued.

This should preserve the current behavior and result in no user-facing changes.

The logs have also been updated which should help with debugging. They surface information like this:

```json
{
  "level": "info",
  "service": "authorize",
  "request-id": "6f645e73-2f59-4ea4-84e6-a4a60d7c4c36",
  "check-request-id": "f11ffdc9-456f-460e-8115-d93949d60801",
  "method": "GET",
  "path": "/",
  "host": "httpbin.localhost.pomerium.io",
  "query": "",
  "session-id": "SOMESESSIONID",
  "allow": true,
  "allowed-reasons": [
    "domain-ok"
  ],
  "deny": false,
  "not-denied-reasons": [
    "valid-client-certificate"
  ],
  "user": "SOMEUSERID",
  "email": "SOMEEMAILADDRESS",
  "databroker_server_version": 2101596356763239000,
  "databroker_record_version": 11,
  "time": "2021-09-17T15:28:17-06:00",
  "message": "authorize check"
}

```

## Related issues
Fixes #2609 

## Checklist
- [x] reference any related issues
- [ ] updated docs
- [x] updated unit tests
- [ ] updated UPGRADING.md
- [x] add appropriate tag (`improvement` / `bug` / etc)
- [x] ready for review

### CODE REVIEW

**Review by desimone — APPROVED** [review]
Being able to get more context on why something was allowed or denied will be super helpful. 

A few small suggestions. 

1. We should probably export custom error types (e.g. `495` and be consistent with http.StatusCodes).
2. I'm not sure how best to markup the logs to indicate the (initially confusing) allow or deny reasons. I'm not sure my suggestion is any better. Open to ideas.

**Review by wasaga — APPROVED** [review]
idk about 

`
"not-denied-reasons":["valid-client-certificate"]
`
it sounds confusing.

File: authorize/check_response.go [hunk 1]
```diff
@@ -23,11 +23,56 @@ import (
...
+	}
+
+	return a.deniedResponse(ctx, in, denyStatusCode, denyStatusText, nil)
+}
+
+func (a *Authorize) handleResultNotAllowed(
+	ctx context.Context,
+	in *envoy_service_auth_v3.CheckRequest,
+	result *evaluator.Result,
+	isForwardAuthVerify bool,
```
[the remark this answers is not in the document]
Comment (calebdoxsey) [reply authorize/check_response.go:61]: Yes. It was pulled out into a variable and is passed as an argument to this function.

**Comment by calebdoxsey** [comment]
I renamed it to `valid-client-certificate-or-none-required`.

**Pull request merged**